tests/benchmarks/test_model_schema_generation.py

```diff
@@ -307,3 +307,115 @@ class PydanticTypeModel(DeferredModel):
         field: field_type
 
     benchmark(rebuild_model, PydanticTypeModel)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_simple_recursive_model_schema_generation(benchmark):
+    class Foo(DeferredModel):
+        a: int = 123
+        sibling: 'Foo' = None
+
+    benchmark(rebuild_model, Foo)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_generic_recursive_model_schema_generation(benchmark):
+    T = TypeVar('T')
+
+    class GenericFoo(DeferredModel, Generic[T]):
+        value: T
+        sibling: Optional['GenericFoo[T]'] = None
+
+    benchmark(rebuild_model, GenericFoo[int])
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_nested_recursive_model_schema_generation(benchmark):
+    class Node(DeferredModel):
+        value: int
+        left: Optional['Node'] = None
+        right: Optional['Node'] = None
+
+    class Tree(DeferredModel):
+        root: Node
+        metadata: Dict[str, 'Tree'] = Field(default_factory=dict)
+
+    benchmark(rebuild_model, Tree)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_nested_recursive_generic_model_schema_generation(benchmark):
+    T = TypeVar('T')
+
+    class GenericNode(DeferredModel, Generic[T]):
+        value: T
+        left: Optional['GenericNode[T]'] = None
+        right: Optional['GenericNode[T]'] = None
+
+    class GenericTree(DeferredModel, Generic[T]):
+        root: GenericNode[T]
+        metadata: Dict[str, 'GenericTree[T]'] = Field(default_factory=dict)
+
+    benchmark(rebuild_model, GenericTree[int])
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_recursive_discriminated_union_with_base_model(benchmark) -> None:
+    class Foo(DeferredModel):
+        type: Literal['foo']
+        x: 'Foobar'
+
+    class Bar(DeferredModel):
+        type: Literal['bar']
+
+    class Foobar(DeferredModel):
+        value: Union[Foo, Bar] = Field(discriminator='type')
+
+    benchmark(rebuild_model, Foobar)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_recursive_discriminated_union_with_typed_dict(benchmark) -> None:
+    class Foo(TypedDict):
+        type: Literal['foo']
+        x: 'Foobar'
+
+    class Bar(TypedDict):
+        type: Literal['bar']
+
+    class Foobar(DeferredModel):
+        value: Union[Foo, Bar] = Field(discriminator='type')
+
+    benchmark(rebuild_model, Foobar)
+
+
+@pytest.mark.benchmark(group='model_schema_generation')
+def test_recursive_discriminated_union_with_pydantic_dataclass(benchmark) -> None:
+    @dataclass
+    class Foo:
+        type: Literal['foo']
+        x: 'Foobar'
+
+    @dataclass
+    class Bar:
+        type: Literal['bar']
+
+    @dataclass(frozen=True, kw_only=True, config={'defer_build': True})
+    class Foobar:
+        value: Union[Foo, Bar] = Field(discriminator='type')
+
+    benchmark(lambda: rebuild_dataclass(Foobar, force=True, _types_namespace={}))
```

Comment:
I think we can remove these ones, I don't expect any schema generation changes with Pydantic models

Reply:
That makes sense. Thanks for the feedback!